src/media-query-provider.js

```diff
@@ -31,10 +33,13 @@ class MediaQueryProvider extends React.Component {
     };
 
     this.mediaQueryListener = this.mediaQueryListener.bind(this);
-  }
 
-  getChildContext() {
-    return this.state;
+    this.mediaRef = this.state.media;
+    this.updateState = _.debounce((newMedia) => {
```

Comment:
why is a `debounce` required?

Reply:
This was a reason for a huge performance issue. I didn’t mean to submit this as part of the pull request. We had the fix on our repo and I forgot that the merge request will take the changes.

In our project, we have an extremely complicated layout with hundreds of components. And we have used the media provider to subscribe for a number of media queries.

The problem is that our media queries overlaps. For example :

When the current size Desktop, all the following media queries will match :

- Min xSmall
- Min Mobile
- Min Tablet
- Min Desktop

We have a lot of these queries (about 16). And for each breakpoint hit, at least 3-4 of them would change.

The way the browser media change listener works is that it would fire the event listener on each of these. Which means for one breakpoint (Desktop to Tablet), the whole view would updated 3-4 times.

In our case, one update for the layout would take about 600ms. For 3-4, that means about 1800-2400ms. Which is so much.

Reply:
The denounce of 50ms, gives the layout events from the browser to accumulate and gets fired at once.

This was a work in progress submitted by an Engineer on my team. I’ll re-review it and let you know. 

Also let me know what do you think about the general concept.

I can also separate this in its own merge request if you like.

Reply:
if you have 16 different breakpoint then you need 16 different listeners regardless

The only issue with having this many breakpoints is that when you manually change the screen size (or do an orientation change on mobile) then it might trigger the listeners quite a few times (and this is where the debounce might be handy) - is this the scenario you mean ?

Reply:
I’ll see also if we can go around the use of debounce, with some sort of custom code

Reply:
Here you're how we're calling the HOC's Provider

```
const queries = {
  largeDesktop: mediaQueryMin('largeDesktop'),
  maxLargeDesktop: mediaQueryMax('largeDesktop'),

  desktop: mediaQueryExact('desktop'),
  minDesktop: mediaQueryMin('desktop'),
  maxDesktop: mediaQueryMax('desktop'),

  tablet: mediaQueryExact('tablet'),
  minTablet: mediaQueryMin('tablet'),
  maxTablet: mediaQueryMax('tablet'),

  mobile: mediaQueryExact('mobile'),
  minMobile: mediaQueryMin('mobile'),
  maxMobile: mediaQueryMax('mobile'),

  minXsmall: mediaQueryMin('xsmall'),
  xsmall: mediaQueryMax('xsmall'),
};

const MediaProvider = props => (
  <MediaQueryProvider queries={queries}>{props.children}</MediaQueryProvider>
);
```

Now, let me walk you throw one breakpoint crossing (Desktop -> Tablet). The following queries will all change their values :


```
  desktop:          true -> false
  minDesktop:   true -> false
  tablet:              false -> true
  maxTablet:      true -> false
```

This would mean 4 refreshes for our whole layout. 400% more time than what's really needed.

The reason we needed to go down this path, is more readability of the comsuming code.

I can't argue though whether or not, this might be a common need among your library consumers. But the one you mentioned is also a viable use case (Dragging the window quickly from very large desktop size, to a small iPad size, would also trigger a number of queries)

But, if you're okay with that, I could try other alternatives and compare the sizes. Using only specialized lib like this : https://bundlephobia.com/result?p=debounce@1.2.0.

In all cases, I'm thinking about separating this discussion in a new issue/pull-request, and see how things goes. Meanwhile, I could apply your comments on this pull-request, to be scheduled in next release.

What do you think ?

Reply:
I get your use case for `debounce` 

Im not really sure what `mediaQueryMin`, `mediaQueryExact` are from - is this some custom code at your end?

Im also a bit confused why you have some duplicate media queries (Eg: 
```
minXsmall: mediaQueryMin('xsmall'),
xsmall: mediaQueryMax('xsmall'),
```)

If you didnt have so many conflicting queries it wouldn't be less of an issue

Having said that im not against adding `debounce` if it will help you - can you raise this mod in a seperate PR ? (its unrelated to React Context api)

Reply:
Done, kindly check our last commit.

The reason we added almost similar queries is to increase the utility for our code base consumer (other developers). Sometimes it's more convenient to use MediaMax, and other times MediaMin would make more sense.